From b46cc6adb06b367da81276a3992c3aea20414629 Mon Sep 17 00:00:00 2001
From: Sidraya Jayagond <sidraya.bj@ti.com>
Date: Mon, 8 Mar 2021 19:32:58 +0530
Subject: [PATCH 1/2] TI IMG video codec driver porting to ti-linux-5.4.y

Porting V4L2 codec driver APIs to latest Kernel 5.4 and
Porting OSAL module for making required compatible changes
to ti-linux-5.4.y.

Signed-off-by: Sidraya Jayagond <sidraya.bj@ti.com>
---
 driver/common/addr_alloc.h |  2 --
 driver/common/hash.c       |  2 --
 driver/common/resource.c   |  2 +-
 driver/common/resource.h   |  2 +-
 driver/decoder/core.c      |  6 +++---
 driver/decoder/decoder.c   | 10 +++++-----
 linux/decoder/vxd_v4l2.c   |  5 +++--
 linux/encoder/Makefile     |  2 +-
 linux/encoder/vxe_v4l2.c   |  5 +++--
 osal/inc/osa_time.h        |  2 +-
 osal/src/linux/osa_time.c  |  4 ++--
 11 files changed, 20 insertions(+), 22 deletions(-)

diff --git a/driver/common/addr_alloc.h b/driver/common/addr_alloc.h
index ae19706..67cf33d 100644
--- a/driver/common/addr_alloc.h
+++ b/driver/common/addr_alloc.h
@@ -20,8 +20,6 @@
 #include "osal/inc/osa_mutex.h"
 #include "ra.h"
 
-#define FALSE 0
-#define TRUE 1
 
 /* Defines whether sequential or random allocation is used */
 enum {
diff --git a/driver/common/hash.c b/driver/common/hash.c
index e3c2048..24116d7 100644
--- a/driver/common/hash.c
+++ b/driver/common/hash.c
@@ -22,8 +22,6 @@
 #include "img_errors.h"
 #include "pool.h"
 
-#define FALSE 0
-#define TRUE 1
 
 /* pool of struct hash objects */
 static struct pool *global_hashpool = NULL;
diff --git a/driver/common/resource.c b/driver/common/resource.c
index 26bd373..3d2a5f3 100644
--- a/driver/common/resource.c
+++ b/driver/common/resource.c
@@ -77,7 +77,7 @@ int32 resource_item_isavailable(uint32 *refcnt)
 /*
  * adds an item (and associated id) to a resource list
  */
-int32 resource_list_add(struct lst_t *list, void *item, uint32 id, uint32 *refcnt)
+int32 resource_list_add_img(struct lst_t *list, void *item, uint32 id, uint32 *refcnt)
 {
 	struct resource_list_elem *listelem = NULL;
 	int32 bfound = 0;
diff --git a/driver/common/resource.h b/driver/common/resource.h
index 540aeb4..278e5c2 100644
--- a/driver/common/resource.h
+++ b/driver/common/resource.h
@@ -28,7 +28,7 @@ int32 resource_item_release(uint32 *refcnt);
 
 int32 resource_item_isavailable(uint32 *refcnt);
 
-int32 resource_list_add(struct lst_t *list, void *item, uint32 id, uint32 *refcnt);
+int32 resource_list_add_img(struct lst_t *list, void *item, uint32 id, uint32 *refcnt);
 
 void *resource_list_pickhead(struct lst_t *list);
 
diff --git a/driver/decoder/core.c b/driver/decoder/core.c
index 29d1383..d7fc0a3 100644
--- a/driver/decoder/core.c
+++ b/driver/decoder/core.c
@@ -2201,7 +2201,7 @@ static int32 core_stream_resource_create(struct core_stream_context *core_str_ct
 		pictres_int->seq_resint = seqres_int;
 #endif
 		/* Add the internal picture resources to the list. */
-		ret = resource_list_add(&core_str_ctx->aux_pict_res_list,
+		ret = resource_list_add_img(&core_str_ctx->aux_pict_res_list,
 					pictres_int, 0, &pictres_int->ref_cnt);
 
 		pictres_int = lst_next(pictres_int);
@@ -2630,7 +2630,7 @@ int32 core_stream_submit_unit(uint32 res_str_id, struct vdecdd_str_unit *str_uni
 	dd_str_context = core_str_ctx->dd_str_ctx;
 	VDEC_ASSERT(dd_str_context);
 
-	ret = resource_list_add(&core_str_ctx->str_unit_list, str_unit, 0,
+	ret = resource_list_add_img(&core_str_ctx->str_unit_list, str_unit, 0,
 				NULL);
 	VDEC_ASSERT(ret == IMG_SUCCESS);
 
@@ -2771,7 +2771,7 @@ int32 core_stream_fill_pictbuf(uint32 buf_map_id)
 	VDEC_ASSERT(ddbuf_map_info->buf_type == VDEC_BUFTYPE_PICTURE);
 
 	/* Add the image buffer to the list */
-	ret = resource_list_add(&core_str_ctx->pict_buf_list, ddbuf_map_info,
+	ret = resource_list_add_img(&core_str_ctx->pict_buf_list, ddbuf_map_info,
 				0, &ddbuf_map_info->ddbuf_info.ref_count);
 
 	return ret;
diff --git a/driver/decoder/decoder.c b/driver/decoder/decoder.c
index 7a391a4..f22f66b 100644
--- a/driver/decoder/decoder.c
+++ b/driver/decoder/decoder.c
@@ -1125,7 +1125,7 @@ decoder_stream_decode_resource_create(struct dec_str_ctx *dec_str_context)
 
 	pict_dec_res->ref_cnt = 1;
 
-	ret = resource_list_add(&dec_str_context->dec_res_lst, pict_dec_res, 0,
+	ret = resource_list_add_img(&dec_str_context->dec_res_lst, pict_dec_res, 0,
 				&pict_dec_res->ref_cnt);
 
 	if (ret != IMG_SUCCESS) {
@@ -2377,7 +2377,7 @@ decoder_picture_decode(struct dec_str_ctx *dec_str_ctx,
 
 		if (resource_item_isavailable(&dec_str_ctx->prev_fe_pict_dec_res->ref_cnt)) {
 			resource_list_remove(&dec_str_ctx->dec_res_lst, dec_str_ctx->prev_fe_pict_dec_res);
-			resource_list_add(&dec_str_ctx->dec_res_lst,
+			resource_list_add_img(&dec_str_ctx->dec_res_lst,
 					  dec_str_ctx->prev_fe_pict_dec_res, 0,
 					  &dec_str_ctx->prev_fe_pict_dec_res->ref_cnt);
 		}
@@ -2553,7 +2553,7 @@ decoder_stream_reference_resource_create(struct dec_str_ctx *dec_str_ctx)
 
 	pict_ref_res->ref_cnt = 1;
 
-	ret = resource_list_add(&dec_str_ctx->ref_res_lst, pict_ref_res, 0,
+	ret = resource_list_add_img(&dec_str_ctx->ref_res_lst, pict_ref_res, 0,
 				&pict_ref_res->ref_cnt);
 	if (ret != IMG_SUCCESS) {
 		OSA_PR_ERR("[USERSID=0x%08X] Failed to add resource",
@@ -4491,7 +4491,7 @@ int32 decoder_service_firmware_response(void *dec_str_ctx_arg, uint32 *msg,
 			if (resource_item_isavailable(&dec_str_ctx->last_be_pict_dec_res->ref_cnt)) {
 				resource_list_remove(&dec_str_ctx->dec_res_lst,
 						     dec_str_ctx->last_be_pict_dec_res);
-				resource_list_add(&dec_str_ctx->dec_res_lst,
+				resource_list_add_img(&dec_str_ctx->dec_res_lst,
 						  dec_str_ctx->last_be_pict_dec_res,
 						  0,
 						  &dec_str_ctx->last_be_pict_dec_res->ref_cnt);
@@ -4512,7 +4512,7 @@ int32 decoder_service_firmware_response(void *dec_str_ctx_arg, uint32 *msg,
 		if (resource_item_isavailable(&dec_pict->cur_pict_dec_res->ref_cnt)) {
 			resource_list_remove(&dec_str_ctx->dec_res_lst,
 					     dec_pict->cur_pict_dec_res);
-			resource_list_add(&dec_str_ctx->dec_res_lst,
+			resource_list_add_img(&dec_str_ctx->dec_res_lst,
 					  dec_pict->cur_pict_dec_res,
 					  0,
 					  &dec_pict->cur_pict_dec_res->ref_cnt);
diff --git a/linux/decoder/vxd_v4l2.c b/linux/decoder/vxd_v4l2.c
index 5f0075a..c2ff49b 100644
--- a/linux/decoder/vxd_v4l2.c
+++ b/linux/decoder/vxd_v4l2.c
@@ -1632,12 +1632,12 @@ static int32 vxd_g_selection(struct file *file, void *fh,
 static const struct v4l2_ioctl_ops vxd_dec_ioctl_ops = {
 	.vidioc_querycap = vxd_dec_querycap,
 
-	.vidioc_enum_fmt_vid_cap_mplane = vxd_dec_enum_fmt,
+	.vidioc_enum_fmt_vid_cap = vxd_dec_enum_fmt,
 	.vidioc_g_fmt_vid_cap_mplane = vxd_dec_g_fmt,
 	.vidioc_try_fmt_vid_cap_mplane = vxd_dec_try_fmt,
 	.vidioc_s_fmt_vid_cap_mplane = vxd_dec_s_fmt,
 
-	.vidioc_enum_fmt_vid_out_mplane = vxd_dec_enum_fmt,
+	.vidioc_enum_fmt_vid_out = vxd_dec_enum_fmt,
 	.vidioc_g_fmt_vid_out_mplane = vxd_dec_g_fmt,
 	.vidioc_try_fmt_vid_out_mplane = vxd_dec_try_fmt,
 	.vidioc_s_fmt_vid_out_mplane = vxd_dec_s_fmt,
@@ -2034,6 +2034,7 @@ static int32 vxd_dec_probe(struct platform_device *pdev)
 	vxd->vfd_dec = vfd;
 	*vfd = vxd_dec_videodev;
 	vfd->v4l2_dev = &vxd->v4l2_dev;
+	vfd->device_caps = V4L2_CAP_VIDEO_M2M_MPLANE | V4L2_CAP_STREAMING;
 	vfd->lock = vxd->mutex;
 
 	video_set_drvdata(vfd, vxd);
diff --git a/linux/encoder/Makefile b/linux/encoder/Makefile
index ddb030d..2a33aca 100644
--- a/linux/encoder/Makefile
+++ b/linux/encoder/Makefile
@@ -75,7 +75,7 @@ vxe-enc-objs += $(OSAL_PATH)/src/linux/osa_reg_io.o
 vxe-enc-objs += $(OSAL_PATH)/src/linux/osa_string.o
 vxe-enc-objs += $(OSAL_PATH)/src/linux/osa_event.o
 
-obj-m += vxe-enc.o
+obj-$(CONFIG_VIDEO_IMG_VXE_ENC) += vxe-enc.o
 
 ccflags-y   += -DOSAL_LINUX
 
diff --git a/linux/encoder/vxe_v4l2.c b/linux/encoder/vxe_v4l2.c
index c79587c..e3d452a 100644
--- a/linux/encoder/vxe_v4l2.c
+++ b/linux/encoder/vxe_v4l2.c
@@ -1602,12 +1602,12 @@ static int32 vxe_s_parm(struct file *file, void *priv,
 static const struct v4l2_ioctl_ops vxe_enc_ioctl_ops = {
 	.vidioc_querycap = vxe_querycap,
 
-	.vidioc_enum_fmt_vid_cap_mplane = vxe_enum_fmt,
+	.vidioc_enum_fmt_vid_cap = vxe_enum_fmt,
 	.vidioc_g_fmt_vid_cap_mplane = vxe_g_fmt,
 	.vidioc_try_fmt_vid_cap_mplane = vxe_try_fmt,
 	.vidioc_s_fmt_vid_cap_mplane = vxe_s_fmt,
 
-	.vidioc_enum_fmt_vid_out_mplane = vxe_enum_fmt,
+	.vidioc_enum_fmt_vid_out = vxe_enum_fmt,
 	.vidioc_g_fmt_vid_out_mplane = vxe_g_fmt,
 	.vidioc_try_fmt_vid_out_mplane = vxe_try_fmt,
 	.vidioc_s_fmt_vid_out_mplane = vxe_s_fmt,
@@ -1840,6 +1840,7 @@ static int32 vxe_enc_probe(struct platform_device *pdev)
 	vxe->vfd = vfd;
 	*vfd = vxe_enc_videodev;
 	vfd->v4l2_dev = &vxe->ti_vxe_dev;
+	vfd->device_caps = V4L2_CAP_VIDEO_M2M_MPLANE | V4L2_CAP_STREAMING;
 	vfd->lock = vxe->mutex;
 
 	video_set_drvdata(vfd, vxe);
diff --git a/osal/inc/osa_time.h b/osal/inc/osa_time.h
index d74fddf..2aee7f6 100644
--- a/osal/inc/osa_time.h
+++ b/osal/inc/osa_time.h
@@ -28,7 +28,7 @@
 #include <linux/time.h>
 
 struct osa_timespec {
-	struct timespec ts;
+	struct timespec64 ts;
 };
 #elif defined(OSAL_SYSBIOS)
 struct osa_timespec {
diff --git a/osal/src/linux/osa_time.c b/osal/src/linux/osa_time.c
index 14958da..82e05b2 100644
--- a/osal/src/linux/osa_time.c
+++ b/osal/src/linux/osa_time.c
@@ -58,9 +58,9 @@ void osa_getnstimeofday(struct osa_timespec *ts_args)
 
 void osa_timespec_sub(struct osa_timespec *end_time, struct osa_timespec *start_time, struct osa_timespec *dif_time)
 {
-    struct timespec *ts = (struct timespec *)(&dif_time->ts);
+	struct timespec64 *ts = (struct timespec64 *)(&dif_time->ts);
 
-    *ts = timespec_sub(*((struct timespec *)(&end_time->ts)), *((struct timespec *)(&start_time->ts)));
+	*ts = timespec64_sub(*((struct timespec64 *)(&end_time->ts)), *((struct timespec64 *)(&start_time->ts)));
 }
 
 /**
-- 
2.17.1

